### PR TITLE
Crystal Support 

### PR DESCRIPTION
**Pull request by nd-02110114**
## DONE
- [x] implement MEGNet Model
- [x] support Material Project Dataset
- [x] refactoring converter method (each model has each converter method)
- [x] migrate CGCNN Model
- [x] confirm test
  - [x] pass `examples/test_example.sh` 
  - [x] pass `pytest -m "not (gpu or slow)" tests`

The detail description is [here](https://github.com/nd-02110114/chainer-chemistry/pull/1)

## TODO
- [ ] add test (WIP)
- [ ] fix path for Material Project Dataset

### CODE REVIEW

File: chainer_chemistry/functions/activation/softplus.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+from chainer import functions
+
+
+def improved_softplus(x):
+    """
```
Comment (corochann) [inline chainer_chemistry/functions/activation/softplus.py]: where is the original source of this function, how did you take `absolute` value??

File: chainer_chemistry/links/readout/cgcnn_readout.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        super(CGCNNReadout, self).__init__()
+        with self.init_scope():
+            self.linear = links.Linear(None, out_dim)
+
+    def __call__(self, atom_feat, atom_idx):
+        average_pool = [functions.mean(atom_feat[idx], axis=0, keepdims=True)
+                        for idx in atom_idx]
+        h = functions.concat(average_pool, axis=0)
+        h = self.linear(h)
+        h = functions.softplus(h)
```
Comment (corochann) [inline chainer_chemistry/links/readout/cgcnn_readout.py]: MEMO: I confirmed original impolementation used normal softplus.
https://github.com/txie-93/cgcnn/blob/master/cgcnn/model.py#L115
(But I wonder shifted_softplus might be better?)

File: chainer_chemistry/links/readout/megnet_readout.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    Args:
+        out_dim (int): dimension of output feature vector
+        in_channels (int): dimension of feature vector associated to
+            each node. Must not be `None`.
+        n_layers (int): number of LSTM layers for set2set
+        processing_steps (int): number of processing for set2set
+        dropout_ratio (float): ratio of dropout
+    """
+
+    def __init__(self, out_dim=32, in_channels=32, n_layers=16,
```
Comment (corochann) [inline chainer_chemistry/links/readout/megnet_readout.py]: n_layers=16 is used in original paper??? I think it is too big?

File: README.md [hunk 4]
```diff
@@ -127,6 +131,7 @@ The following datasets are currently supported:
 - Tox21 [9]
 - MoleculeNet [11]
 - ZINC (only 250k dataset) [12, 13]
+- Material Project (日付を入れる) [23]
```
Comment (corochann) [inline README.md]: TODO later

File: chainer_chemistry/functions/activation/softplus.py (hunk 1 above)
Comment (nd-02110114) [reply chainer_chemistry/functions/activation/softplus.py]: Source : https://github.com/materialsvirtuallab/megnet/blob/f91773f0f3fa8402b494638af9ef2ed2807fcba7/megnet/activations.py#L6

**Comment by nd-02110114** [comment]
I wrote tests for CGCNN & MEGNet. Currently, all tests pass the forward test, 
but doesn't pass the backward test except `test_megnet_readout.py`.

Now, most of the backward test are commented out and 
I investigate why the backward test fails.

File: chainer_chemistry/dataset/preprocessors/megnet_preprocessor.py [hunk 5]
```diff
@@ -0,0 +1,422 @@
+import os
```
Comment (corochann) [inline chainer_chemistry/dataset/preprocessors/megnet_preprocessor.py]: Did you refer some code (original impelmentation)? If you copy some methods from some repository, I think it is better to add reference link like below.
https://github.com/pfnet-research/graph-nvp/blob/master/utils/molecular_metrics.py#L2

File: chainer_chemistry/models/megnet.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
+import chainer
+from chainer import functions
+
+
+from chainer_chemistry.links.update.megnet_update import MEGNetUpdate
+from chainer_chemistry.links.readout.megnet_readout import MEGNetReadout
+
+
+def reshaped_feat(feat, idx):
+    max_idx = max(idx)
```
Comment (corochann) [inline chainer_chemistry/models/megnet.py]: Change below, to work on GPU
```
    xp = get_array_module(idx)
    max_idx = int(xp.max(idx))
```

File: chainer_chemistry/models/megnet.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
+import chainer
+from chainer import functions
+
```
Comment (corochann) [inline chainer_chemistry/models/megnet.py]: add `from chainer.backend import get_array_module`

File: examples/mp/README.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+To train a model, run the following:
+
+On the CPU:
+```angular2html
+python train_mp.py
+```
+
+On the GPU:
+```angular2html
+python train_mp.py -g 0
```
Comment (corochann) [inline examples/mp/README.md]: --device

File: examples/mp/README.md [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+The perform inference using a pretrained model, run the following:
+
+On the CPU:
+```
+python predict_mp.py [-i /path/to/training/result/directory]
+```
+
+On the GPU:
+```
+python predict_mp.py -g 0 [-i /path/to/training/result/directory]
```
Comment (corochann) [inline examples/mp/README.md]: --device

**Pull request merged**